src/lib/cache/util.go

```diff
@@ -0,0 +1,73 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"errors"
+	"reflect"
+	"sync"
+)
+
+func indirect(reflectValue reflect.Value) reflect.Value {
+	for reflectValue.Kind() == reflect.Ptr {
+		reflectValue = reflectValue.Elem()
+	}
+	return reflectValue
+}
+
+func simpleCopy(dst interface{}, src interface{}) error {
+	from := indirect(reflect.ValueOf(src))
+	to := indirect(reflect.ValueOf(dst))
+	if !to.CanAddr() {
+		return errors.New("dst value is unaddressable")
+	}
+
+	if !from.Type().ConvertibleTo(to.Type()) {
+		return errors.New("src value is not convertible to the dst value")
+	}
+
+	to.Set(from.Convert(to.Type()))
+
+	return nil
+}
+
+type keyMutex struct {
+	m *sync.Map
+}
+
+func (km keyMutex) Lock(key interface{}) {
+	m := sync.Mutex{}
+	act, _ := km.m.LoadOrStore(key, &m)
+
+	mm := act.(*sync.Mutex)
+	mm.Lock()
+	if mm != &m {
+		mm.Unlock()
+		km.Lock(key)
+		return
+	}
+
+	return
+}
+
+func (km keyMutex) Unlock(key interface{}) {
+	act, exist := km.m.Load(key)
+	if !exist {
+		panic("unlock of unlocked mutex")
+	}
+	m := act.(*sync.Mutex)
+	km.m.Delete(key)
```

Comment:
I don't think you have to delete everytime, or you may use: https://godoc.org/sync#Map.LoadAndDelete

Reply:
> LoadAndDelete deletes the value for a key, returning the previous value if any. The loaded result reports whether the key was present.

`LoadAndDelete` is not suitable here because it returns the `previous` value, the `Unlock` mutex must be the current one, not the `previous`.

Delete the mutex every time here to ensure that `act, _ := km.m.LoadOrStore(key, &m)` always let `act` be `m` when the `key` is not locked in the `Lock` method, otherwise, the `Lock` method will go to the `if mm != &m {` block which may drop in an infinite loop.

Reply:
Current code works, but I believe the `previous` here means the `current` in your context.
It's the value that were mapped to the key before it's deleted.